### PR TITLE
feat(system): add native Termux (Android) support

### LINKED ISSUE #276: feat(system): add native Termux (Android) support
## Pre-flight

- [x] I searched existing issues and found no duplicate
- [x] I understand that a maintainer must add `status:approved` before a PR can be opened

## Affected Area

`System Detection` · `Installation Pipeline`

## 🇪🇸 Motivación — Por qué esto importa

Imaginate esto: estás en la calle, sin portátil, pero con tu móvil Android y Termux instalado. Tienes tokens de Claude disponibles y una idea que no puede esperar. ¿Qué haces? Pues resulta que **no podías usar `gga`**, porque `gentle-ai` no era compatible con Android/Termux.

Esto va de tener el mejor agente de IA de desarrollo **incluso en el bolsillo**. Porque nunca se sabe cuándo vas a tener tokens de Claude de nuevo... hay que estar preparado 😄

La realidad es que Termux convierte un teléfono Android en un entorno de desarrollo Linux real, y no hay razón para que `gentle-ai` no funcione ahí. Ahora funciona. Probado en un dispositivo real con Termux.

---

## 🇬🇧 Problem Statement

`gentle-ai` does not support Android (Termux) as a platform. Running `gga` on Termux fails at multiple points:

1. **OS Detection**: `runtime.GOOS` reports `"android"` on Termux, which is not in the supported OS list → immediate rejection with "unsupported operating system".
2. **Path Resolution**: Standard Unix paths (`/usr/bin`, `/etc`, `/tmp`) don't exist on Android. Termux maps them under `$PREFIX` (typically `/data/data/com.termux/files/usr`).
3. **Binary Downloads**: Pre-built Linux binaries (glibc) are incompatible with Android's Bionic libc → `engram` binary crashes with `SIGILL` or linker errors.
4. **Sudo Usage**: `sudo` doesn't exist in Termux — any command using `sudo` fails.
5. **PATH Persistence**: Adding to `PATH` via shell rc files requires detecting the correct shell config file in Termux.

### System Detection Layer
- Add `"android"` to `IsSupportedOS()` as a first-class supported OS
- Map `GOOS=android` → `LinuxDistro=termux`, `PackageManager=apt`, `Supported=true` in `resolvePlatformProfile()`
- Update error messages to include Android in the supported OS list

### Path Resolution — `PathResolver` Interface
- Introduce `PathResolver` interface with `DefaultResolver` (pass-through) and `TermuxResolver`
- `TermuxResolver` maps `/usr/*` → `$PREFIX/*`, `/bin/*` → `$PREFIX/bin/*`, `/etc/*` → `$PREFIX/etc/*`, `/tmp/*` → `$PREFIX/tmp/*`
- `NewResolverForDistro()` factory returns the appropriate resolver based on detected distro

### Binary Installation
- For `engram` on Android: compile from source via `go install` with `-ldflags=-extldflags=-pie` (Bionic requires Position Independent Executables)
- For `go install` upgrades: inject PIE ldflags when `LinuxDistro == "termux"` or `GOOS == "android"`

### Sudo Suppression
- Skip `sudo` prefix for `apt-get`, `pacman`, `dnf` commands when `LinuxDistro == "termux"`

### PATH Persistence
- Detect Termux via `TERMUX_VERSION` env var
- Persist PATH additions to `~/.bashrc` or `~/.zshrc` based on `$SHELL`

## Alternatives Considered

- **proot/chroot**: Run a full Linux distribution inside Termux. Rejected — adds complexity and defeats the purpose of native Android support.
- **Cross-compile with Bionic**: Build binaries targeting Android specifically. Rejected — too much CI complexity for a niche platform; `go install` from source is simpler and works.

## Additional Context

- Tested and validated on a real Android device running Termux
- All existing tests continue to pass (the only failures are pre-existing Windows `echo` issues unrelated to this change)
- New test coverage added for Android OS detection, guard messages, and path resolution

### PR DESCRIPTION
**Pull request by Snakeblack**
## Linked Issue

Closes #276

## PR Type

- [x] `type:feature` - New feature (non-breaking change that adds functionality)

## Summary

Adds native Android/Termux support to `gentle-ai`:

- Detects Android/Termux as a first-class platform profile through `GOOS=android`.
- Resolves Termux paths through a dedicated resolver, including `$PREFIX` aware paths.
- Keeps Android installs and upgrades on source compilation with PIE flags, avoiding incompatible glibc release binaries.
- Preserves current `main` hardening for checksum verification, Windows/Scoop docs, Go availability detection, and pinned GGA upgrade behavior.

## Changes

| Area | Files | Change |
|------|-------|--------|
| Platform detection | `internal/system/detect.go`, tests | Adds Android as a supported OS and maps it to the Termux profile. |
| Path resolution | `internal/system/resolver.go`, `internal/system/path.go`, tests | Adds prefix-aware path resolution and Termux shell PATH persistence. |
| Install commands | `internal/installcmd/resolver.go`, tests | Centralizes sudo/rootless behavior and keeps Android commands sudo-free. |
| Engram install | `internal/components/engram/download.go`, tests | Uses versioned `go install` with PIE flags on Android while preserving checksum verification for binary downloads. |
| Upgrade strategy | `internal/update/upgrade/strategy.go`, tests | Applies PIE flags for Android Go installs and preserves current main upgrade hardening. |
| Installer script | `scripts/install.sh` | Forces Android/Termux to use `go install`; blocks brew/binary methods on Android. |
| Docs/specs | `docs/platforms.md`, `openspec/...` | Documents Android/Termux support and the review path. |

## Key Technical Decisions

1. Android/Termux is detected through `GOOS=android`, not through Linux `ID=termux` or `TERMUX_VERSION` in Go code. This avoids a half-detected `linux + termux` state.
2. Release binaries remain for `linux`, `darwin`, and `windows`. Android uses source compilation because Linux glibc binaries are incompatible with Android Bionic libc.
3. Android builds and upgrades use `-ldflags=-extldflags=-pie`, required by Android executable loading rules.
4. Rootless Termux flows do not use `sudo`.
5. The path resolver keeps platform-specific path mapping isolated from install and update logic.

## Review workload note

This PR exceeds the current 400-line review budget, but it predates the chained-PR policy and keeps one cohesive platform-support unit together: Android/Termux detection, prefix-aware paths, install behavior, upgrade behavior, tests, and docs.

Splitting it now would add review churn and risk losing the existing validation and context. Treating this as a maintainer-approved `size:exception` is the lower-risk path for this already-open PR.

To keep review focused, please review in this order:

1. `internal/system/*` platform detection and path resolution
2. `internal/installcmd/*` sudo/rootless install behavior
3. `internal/components/engram/*` Android source install with PIE
4. `internal/update/upgrade/*` upgrade strategy
5. docs/OpenSpec updates

## Test Plan

- [x] `go test ./internal/system ./internal/installcmd ./internal/components/engram ./internal/update/upgrade`
- [x] `git diff --check`
- [x] Manual validation on a physical Android device running Termux: detection, install, upgrade, PATH persistence, and PIE source builds.
- [x] PR branch is mergeable against current `main` after conflict resolution.
- [ ] Full CI validates broad repository checks on GitHub runners.

## Contributor Checklist

- [x] Linked an approved issue (`Closes #276`)
- [x] Added exactly one `type:*` label (`type:feature`)
- [x] Relevant unit tests pass
- [x] Documentation updated
- [x] Conventional commit format used
- [x] No `Co-Authored-By` trailers

### CODE REVIEW

File: internal/components/engram/download.go [hunk 1]
```diff
@@ -37,6 +38,10 @@ var (
 // This is the non-brew installation method for Linux and Windows.
 // On macOS, brew handles engram transitively and this should not be called.
 func DownloadLatestBinary(profile system.PlatformProfile) (string, error) {
+	if profile.OS == "android" {
+		return installViaGo(profile)
+	}
+
```
[the remark this answers is not in the document]
Comment (Snakeblack) [reply internal/components/engram/download.go:52]: Valid observation about testability. However, this follows the same pattern used throughout the codebase (e.g., `engramInstallDir`, `installViaGo` in the same file). Extracting a command runner interface for a single call site would be over-engineering at this stage.

Current coverage:
- `installViaGo` is tested **manually on a real Termux device** — end-to-end `go install` with PIE flags, binary execution, and PATH resolution all confirmed working
- The PIE flag logic IS tested in `TestRunStrategy_GoInstallUpgradeAndroid` (via `strategy.go`'s `goInstallUpgrade`, which uses the same pattern with an injectable `execCommand`)

If `installViaGo` grows in complexity or gains more callers, injecting the command runner makes sense as a follow-up. For now, the risk is low and coverage is adequate.

File: internal/system/resolver.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	if !strings.HasPrefix(path, "/") {
+		return path
+	}
+
+	// Handle /usr, /bin, /etc prefixes for Termux layout.
+	if strings.HasPrefix(path, "/usr") {
+		return filepath.Join(r.Prefix, strings.TrimPrefix(path, "/usr"))
+	}
+	if strings.HasPrefix(path, "/bin") {
+		return filepath.Join(r.Prefix, "bin", strings.TrimPrefix(path, "/bin"))
```
[the remark this answers is not in the document]
Comment (Snakeblack) [reply internal/system/resolver.go]: This is a false positive. On Unix/Linux (the only platforms where `TermuxResolver` runs), `filepath.Join("/data/data/com.termux/files/usr", "/bin/bash")` correctly produces `/data/data/com.termux/files/usr/bin/bash`. Go's `filepath.Join` does NOT discard the prefix when both arguments are absolute on Unix — it concatenates them. This is only an issue on Windows, where `TermuxResolver` is never used.

**Evidence:**
- All `TestTermuxResolver_Resolve` test cases pass on both Windows and Linux CI
- Manually validated on a real Android device with Termux — all paths resolve correctly
- The `filepath.ToSlash()` normalization in tests confirms cross-platform correctness

No change needed.

File: e2e/docker-test.sh [hunk 3]
```diff
@@ -21,6 +21,16 @@ YELLOW='\033[1;33m'
...
 
+# ---------------------------------------------------------------------------
+# Pre-flight: Docker must be available
+# ---------------------------------------------------------------------------
+if ! command -v docker &>/dev/null; then
+    printf "${RED}[ORCH]${NC} docker is not installed or not in PATH.\n"
+    printf "${RED}[ORCH]${NC} E2E tests require Docker. On platforms without Docker support\n"
+    printf "${RED}[ORCH]${NC} (e.g. Android/Termux), these tests cannot run.\n"
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (Snakeblack) [reply e2e/docker-test.sh:32]: Intentionally kept as `exit 1`. Rationale:

- `exit 0` would make CI pipelines silently pass when Docker is missing due to misconfiguration — masking real problems
- `exit 2` (skip) would require all CI consumers to handle a non-standard exit code
- The descriptive error message already makes it clear WHY it failed — the user/CI can distinguish "Docker not installed" from "tests failed"
- On Termux specifically, users won't run `docker-test.sh` expecting it to work — Docker cannot run on Android (no kernel namespace/cgroup support)

If a dedicated "skip" exit code is desired in the future, it can be added as a follow-up without breaking current behavior.

File: internal/system/resolver.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	if strings.HasPrefix(path, "/etc") {
+		return filepath.Join(r.Prefix, "etc", strings.TrimPrefix(path, "/etc"))
+	}
+	if strings.HasPrefix(path, "/tmp") {
+		return filepath.Join(r.Prefix, "tmp", strings.TrimPrefix(path, "/tmp"))
```
[the remark this answers is not in the document]
Comment (Snakeblack) [reply internal/system/resolver.go:52]: Duplicate of the thread at line 44 — same `filepath.Join` concern. See reply there: this is a false positive on Unix. `filepath.Join` concatenates correctly when both args are absolute paths on Linux/Android. All tests pass and manually validated on real Termux device.

**Review by Alan-TheGentleman — CHANGES_REQUESTED** [review]
## Review: feat(system): add native Termux (Android) support

Solid contribution — the architecture is clean (PathResolver interface, clean detection branching, PIE awareness), the test coverage is good, and the SDD artifacts provide excellent context. Two issues need attention before merge:

### Must Fix

**1. Inconsistent Termux detection between `GOOS=android` and `GOOS=linux` + `ID=termux`**

The `case "android"` branch in `resolvePlatformProfile` correctly sets `LinuxDistro=termux`, but the PR also adds Termux detection in the `"linux"` branch via `detectLinuxDistro`. If someone runs on `GOOS=linux` with `ID=termux` in os-release (which Termux does report), the profile gets `OS=linux` + `LinuxDistro=termux`. But downstream checks in `download.go` (`profile.OS == "android"`) and `engramInstallDir` won't trigger, creating a half-detected state.

Recommendation: Remove the Termux case from `detectLinuxDistro` and the `LinuxDistroTermux` entry in the linux distro switch — `GOOS=android` handles the full detection. Or, update all downstream checks to also match on `LinuxDistro == "termux"`.

**2. Dead code in `engramAssetURL` (android→linux mapping)**

`DownloadLatestBinary` returns early for `OS=android` via `installViaGo`, so the `goos == "android"` mapping in `engramAssetURL` is unreachable. If defensive, add a comment. If not needed, remove it.

### Suggestions

- `persistPathTermux`: the `\n` prefix creates a blank first line when writing to a new RC file
- `installViaGo`: uses `os.Getenv` directly instead of the testable `osGetenv` pattern used elsewhere
- Sudo suppression in `ResolveDependencyInstall`: the prepend logic is duplicated 3 times — consider a small helper
- `echo ok` → `go version` in unrelated tests: worth mentioning in the PR description as an incidental fix

Great work on the PathResolver abstraction and the thorough test coverage. Looking forward to the updated version!

**Comment by Snakeblack** [comment]
Hey Alan, thanks for the thorough review — every point was valid. Here's what we addressed:

---

## Must Fix 1 — Inconsistent Termux detection (`GOOS=android` vs `GOOS=linux + ID=termux`)

**Resolved in `06413cd`.** We chose your first recommendation: remove Termux from the Linux detection path entirely.

- Removed `LinuxDistroTermux` from the `case "linux"` distro switch in `resolvePlatformProfile`
- Removed `if id == LinuxDistroTermux` from `detectLinuxDistro`
- `GOOS=android` is now the **exclusive** Termux detection path — no half-detected state possible
- Added an explicit comment in `detectLinuxDistro` documenting this decision
- Updated tests: distro matrix now expects `unknown` for `ID=termux`, PIE test uses `OS:"android"`

## Must Fix 2 — Dead code in `engramAssetURL` (`android→linux` mapping)

**Resolved in `06413cd`.** Removed the unreachable `if goos == "android" { goos = "linux" }` block. Added a comment explaining that Android never reaches this function since `DownloadLatestBinary` returns early via `installViaGo()`.

---

## Suggestions

### `persistPathTermux` — `\n` prefix on new rc files
**Fixed in `06413cd`.** The `exportCmd` now conditionally omits the leading `\n` when writing to an empty file.

### Sudo suppression duplication in `ResolveDependencyInstall`
**Fixed in `06413cd`.** Extracted a `withSudo()` helper that prepends `"sudo"` unless the profile indicates Termux. Replaces the 3 duplicate if-prepend blocks.

### `installViaGo` — `os.Getenv` vs testable `osGetenv`
**Not applied.** This is a single call in a function that's already tested via `exec.Command` mocking. Adding `osGetenv` indirection here felt like over-engineering for one call site. Happy to revisit if you feel strongly about it.

### `echo ok` → `go version` in unrelated tests
**Pre-existing** — this was an incidental fix in an earlier commit to make tests pass on environments where `echo` isn't available as a standalone binary (Windows). Noted for PR description clarity.

---

## Additional hardening (scripts + Copilot review)

After your review, we also:

- **`scripts/install.sh`** (`deff64b`): Set `GORELEASER_OS=""` for Android (no release assets exist), added 4 layers of guards to make binary download on Android impossible (auto-detect fatal, force-method fatal, empty GORELEASER_OS fatal, install_binary guard). Preserved your original priority comments — only appended the Android exception note.

- **`TermuxResolver` prefix matching** (`145091f`): Tightened from `HasPrefix("/usr")` → `path == "/usr" || HasPrefix("/usr/")` to prevent false matches on paths like `/usrbin` or `/etcetera`. Added boundary edge-case tests.

- **Shell injection guard** (`145091f`): Added character validation in `persistPathTermux` rejecting backticks, quotes, `$`, and newlines before writing to rc files.

- **Error message improvement** (`145091f`): `goInstallUpgrade` error now includes full args (including PIE flags) for easier Termux debugging.

- **`engramInstallDir` android branch** (`145091f`): Documented as defensive/future-proofing in docstring, added test case.

---

**All changes have been validated on a physical Android device running Termux** — detection, `go install` with PIE flags, path resolution, and the install script flow. Tests pass on both Windows and Linux (the remaining failures in CI are pre-existing Windows-specific issues unrelated to this PR).

Let me know if anything else needs adjustment!

**Review by Snakeblack — COMMENTED** [review]
Hice los cambios que me propusiste hace unos días, pero no se integro el feature a main

**Review by Alan-TheGentleman — CHANGES_REQUESTED** [review]
Termux support is a real product expansion, not just a small platform patch.

Please move this through a design issue first or split it into smaller approved slices: platform detection, path resolution, install behavior, upgrade behavior, and docs. The current PR is too broad to review safely as one unit.

**Comment by Alan-TheGentleman** [comment]
Heads-up: this PR is now showing a merge conflict against `main`. We shipped `v1.31.0` (and a number of follow-up fixes) over the past two days — several files this PR touches were also modified upstream.

**Could you rebase against current `main` and resolve the conflicts?**

```bash
git fetch upstream main
git rebase upstream/main
# resolve any conflicts
git push --force-with-lease
```

Once it's green again I can review for merge. Sorry for the drift — it's the cost of the cleanup velocity. Thanks for sticking with it.

**Comment by Snakeblack** [comment]
Hey Alan, thanks for the heads-up.

The branch is updated against current `main`, the merge conflicts are resolved, and the PR is now showing as mergeable. I also rebuilt the PR description with clean Markdown and added the explicit `size:exception` rationale, since this PR predates the current 400-line/chained-PR policy and the Termux work is still one cohesive platform-support unit.

Validated with:

- `go test ./internal/system ./internal/installcmd ./internal/components/engram ./internal/update/upgrade`
- `git diff --check`
- PR mergeability check: `MERGEABLE`

Could you take another look when you have a chance?

**Comment by Alan-TheGentleman** [comment]
Thanks for the rebase and for addressing the earlier review points. I did another full pass.

The core Termux direction looks good, but `main` moved again after #257 and this PR is now conflicting. Because `maintainerCanModify` is disabled on the fork, I cannot push the fix directly here. I prepared a maintainer validation branch/PR with the conflict resolution and the review fixes here: #765.

Important fix included in that branch: Android `gentle-ai` self-upgrade now routes to versioned `go install` with PIE flags instead of trying to download a non-existent Android release asset. It also keeps standard Linux self-upgrade on release binaries, even when Go is installed.

Validation run on the resolved branch:

- `bash -n scripts/install.sh e2e/docker-test.sh`
- `shellcheck scripts/install.sh e2e/docker-test.sh`
- `go test ./internal/update ./internal/update/upgrade ./internal/system ./internal/installcmd ./internal/components/engram`
- `go test ./...`

All passed locally. The remaining policy issue is size: this is now ~1.1k changed lines, so we either need the maintainer-approved `size:exception` path or a split. Given this PR predates the current 400-line policy and the feature is a cohesive platform-support unit, I think `size:exception` is reasonable if maintainers agree.

**Comment by Snakeblack** [comment]
Hi Alan, thanks for preparing #765 and calling out the upgrade edge case.

I updated #277 with your validation branch resolution, merged current `main`, and pushed one extra fix on top: `scripts/install.sh` now uses the latest release tag as a versioned `go install` target instead of `@latest`, while preserving Android PIE flags.

Validation:

- `bash -n scripts/install.sh e2e/docker-test.sh` (run from an LF checkout)
- `git diff --check origin/main...HEAD`
- `go test ./internal/system ./internal/installcmd ./internal/components/engram ./internal/update/upgrade`
- Fresh focused review of the installer fix: approved, no blocking issues

PR #277 is now `MERGEABLE` at `2de32b5`. I also tried to add the `size:exception` label, but the fork account does not have permission to apply upstream labels. The PR body still documents the size-exception rationale.

**Comment by Alan-TheGentleman** [comment]
Closing this in favor of #765.\n\nThe Termux work still matters, but #765 supersedes this branch with maintainer conflict cleanup and broader install/update/test coverage.
